pkg/controller/application/app_controller.go

```diff
@@ -0,0 +1,295 @@
+/*
+ Licensed to the Apache Software Foundation (ASF) under one
+ or more contributor license agreements.  See the NOTICE file
+ distributed with this work for additional information
+ regarding copyright ownership.  The ASF licenses this file
+ to you under the Apache License, Version 2.0 (the
+ "License"); you may not use this file except in compliance
+ with the License.  You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+package application
+
+import (
+	"os"
+	"time"
+
+	"k8s.io/apimachinery/pkg/api/equality"
+
+	"go.uber.org/zap"
+	apiextensionsclient "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/clientcmd"
+
+	corecache "github.com/apache/incubator-yunikorn-core/pkg/cache"
+	appv1 "github.com/apache/incubator-yunikorn-k8shim/pkg/apis/yunikorn.apache.org/v1alpha1"
+	"github.com/apache/incubator-yunikorn-k8shim/pkg/appmgmt/interfaces"
+	shimcache "github.com/apache/incubator-yunikorn-k8shim/pkg/cache"
+	applicationclient "github.com/apache/incubator-yunikorn-k8shim/pkg/client/clientset/versioned"
+	appscheme "github.com/apache/incubator-yunikorn-k8shim/pkg/client/clientset/versioned/scheme"
+	appinformers "github.com/apache/incubator-yunikorn-k8shim/pkg/client/informers/externalversions"
+	applister "github.com/apache/incubator-yunikorn-k8shim/pkg/client/listers/yunikorn.apache.org/v1alpha1"
+	"github.com/apache/incubator-yunikorn-k8shim/pkg/common/constants"
+	"github.com/apache/incubator-yunikorn-k8shim/pkg/common/events"
+	"github.com/apache/incubator-yunikorn-k8shim/pkg/log"
+)
+
+type AppManager struct {
+	amProtocol interfaces.ApplicationManagementProtocol
+	controller *Controller
+}
+
+func NewAppManager(amProtocol interfaces.ApplicationManagementProtocol) *AppManager {
+	return &AppManager{
+		amProtocol: amProtocol,
+	}
+}
+
+// this implements AppManagementService interface
+func (appMgr *AppManager) Name() string {
+	return constants.AppManagerHandlerName
+}
+
+// this implements AppManagementService interface
+func (appMgr *AppManager) ServiceInit() error {
+	return nil
+}
+
+// this implements AppManagementService interface
+func (appMgr *AppManager) Start() error {
+	appMgr.controller = NewController(appMgr)
+	appMgr.controller.Run()
+	return nil
+}
+
+// this implements AppManagementService interface
+func (appMgr *AppManager) Stop() {
+	// noop
```

Comment:
this is not a noop. it needs to stop the controller

Reply:
Can you please elaborate on what exactly do you expect about stopping the controller? Closing the channel, or something else?